### ISSUE TITLE
Pressing Control + Super inserts ^S

### ISSUE THREAD

**Issue by beelzebielsk** [equalsraf/neovim-qt#593]
When I press the Control and Super key simultaneously, it is as if ^S (Control + S) was sent to neovim-qt.

I'm using version 0.2.13 on arch linux. I tried this both with my normal rc file and with no rc file whatsoever. Same behavior.

**Comment by jgehrig** [comment]
I cannot reproduce this behavior... Could you please elaborate?

To make sure we are on the same page... I assume Super/Meta/Windows are all the same key? :)

**Example:**
1. Using KDE/Plasma with nvim-qt (version 0.2.12; nvim 0.3.7).
2. Open nvim-qt with command `nvim-qt -- -u NORC`.
3. In `Insert` mode, press `<Ctrl + Windows Key>`.
4. Notice nothing is inserted.
5. Map `<C-S>` so we can observe when `<C-S>` is pressed, `:map <C-S> :ehco('<C-S>' pressed!')<CR>`
6. Press `<Ctrl + Windows Key>`.
7. Notice no message displays.

**Comment by beelzebielsk** [comment]
Hi, thanks for getting back to me.

I believe they are all the same key, yes. Super is the Windows key on my keyboard.

1. Using i3 with nvim-qt (version 0.2.14, nvim 0.4.2).
2. Open nvim-qt command `nvim-qt -- -u NORC`.
3. In `Insert` mode, press <Ctrel + Windows Key>.
4. The characters `^S` appear in the buffer.

That's what happens for me.

I'm on Arch Linux if that helps.

**Comment by jgehrig** [comment]
No worries, I'm glad to help!

> I believe they are all the same key, yes. Super is the Windows key on my keyboard.

Everyone has a different name for that key :)

> 1. Using i3 with nvim-qt (version 0.2.14, nvim 0.4.2).
> 2. Open nvim-qt command nvim-qt -- -u NORC.
> 3. In Insert mode, press <Ctrl + Windows Key>.
> 4. The character (characters?) ^S appear in the buffer.

I do not see the `^S` behavior for these steps. Perhaps Desktop Manager (KDE vs i3) has an impact?

**Some Debugging Ideas:**
Do you have the ability to easily boot into another Desktop Manager? If so, what happens?

Try building with the following `Shell::keyPressEvent()` patch applied:
``` patch
diff -r 5556b921b590 src/gui/shell.cpp
--- a/src/gui/shell.cpp Wed Sep 18 18:06:11 2019 -0400
+++ b/src/gui/shell.cpp Mon Sep 30 23:35:11 2019 -0400
@@ -1076,6 +1076,8 @@
                return;
        }
 
+       qDebug() << ev;
+
        m_nvim->api0()->vim_input(m_nvim->encode(inp));
        // FIXME: bytes might not be written, and need to be buffered
 }
```

I see these messages when I follow your steps (patch applied):
```
( "Received unknown redraw notification ..." can be ignored/omitted )
QKeyEvent(KeyPress, Key_I, text="i")
```
Notice `<Ctrl + Super>` is omitted, even though these keys are pressed. What do you see?

On my system, another key is needed for the event to send, such at `<Ctrl + Super + V>`:
`QKeyEvent(KeyPress, Key_V, ControlModifier|MetaModifier, text="\u0016")`

**Comment by beelzebielsk** [comment]
I have not tried the debugging idea yet, but I did try using nvim-qt in two different display managers: XFCE4 and KDE. I had the same problem in XFCE4. Different in KDE.

In KDE, when pressing `<C-Super>` in insert mode, I would not see the `^S` character show up, though i would see the cursor disappear until I pressed another character, like a or i. However, when I was typing in command-line mode and pressed `<C-Super>` I see `^S` come up in the command-line.

So try entering command-line mode (ie pressing `:`).

**Comment by beelzebielsk** [comment]
I have also had trouble reproducing the original behavior since installing KDE, and even after unintalling all the KDE packages and their dependencies. Though I can still reproduce the `^S` showing up on vim command-line.

**Comment by beelzebielsk** [comment]
I tried applying the patch using `git apply`, I had no luck. I am not sure why. Kept telling me something like 'line 12 of patch is corrupt'.

I added the extra two lines manually, though. Then build neovim-qt using the directions from the README, then navigated to `build/bin` and ran `./nvim-qt` from my terminal. I expected to see messages in standard output there, but saw nothing printed. Where would I see these messages you're talking about?

**Comment by beelzebielsk** [comment]
I have a second computer with neovim-qt installed. i3, archlinux, neovim-qt version 2.14, neovim version 0.4.2, Qt version 5.13.1. It does not have KDE installed. I cannot reproduce the behavior there either, where I enter '^S' in insert mode. But I can enter it in command-line mode on this computer.

So for now, ignore '^S' coming up in insert mode, since I can no longer reproduce it and don't know why. Just look at '^S' coming up while in command-line mode.

1. Using i3 with nvim-qt (version 0.2.14, nvim 0.4.2).
2. Open nvim-qt command nvim-qt -- -u NORC.
3. Press `:` to enter command-line mode.
4. In command-line mode, press <Ctrl + Windows Key>.
5. The character (characters?) `^S` appear in the command buffer.

**Comment by jgehrig** [comment]
I'm not sure I follow what has been tested...

To display `qDebug()` messages, you will need to do a Debug build. Here are commands:
```
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug ..
make
```

No worries, adding the extra line manually is fine. The goal is to monitor the key events, and how they are passed through Qt to neovim. The purpose of the testing is to eliminate possible causes of error (Qt, Window Manager, neovim-qt).

1. Does this happen in KDE?
2. Does this happen in XFCE?
3. Does this happen in i3?

Detailed numbered steps down to the keystroke are helpful, especially if you can show the associated debug messages.

I run Linux/KDE; if you have a KDE case for this bug, that would be useful for me...

**Comment by beelzebielsk** [comment]
I apologize for the confusing comments earlier. I was confused, having more trouble than I'd expected. The main confusing thing was that I didn't remember having to use `-u NORC` to reproduce the bug, so it suddenly looked like my bug disappeared inexclipably.

I follow these exact directions each time:

1. Run one of the display managers (i3, KDE, XFCE4).
2. Run debug build of neovim-qt `./nvim-qt -- -u NORC`.
3. Press `i<c-Super><esc>ZQ` to show a `^S` and then quit Neovim-qt.

Below are the debug messages after following the above process in each desktop manager. They're all identical, so the desktop manager doesn't make a difference. In each, where it says keypress `Key_Super_L` is when the `^S` appears in the buffer.

### In KDE

```
Neovim API version compatible with 0 supported 5
Received unknown option "ext_hlstate" QVariant(bool, false)
QKeyEvent(KeyPress, Key_I, text="i")
QKeyEvent(KeyPress, Key_Super_L, ControlModifier)
QKeyEvent(KeyPress, Key_Escape, text="\u001B")
QKeyEvent(KeyPress, Key_Z, ShiftModifier, text="Z")
QKeyEvent(KeyPress, Key_Q, ShiftModifier, text="Q"
```

### XFCE4

```
Neovim API version compatible with 0 supported 5
Received unknown option "ext_hlstate" QVariant(bool, false)
QKeyEvent(KeyPress, Key_I, text="i")
QKeyEvent(KeyPress, Key_Super_L, ControlModifier)
QKeyEvent(KeyPress, Key_Escape, text="\u001B")
QKeyEvent(KeyPress, Key_Z, ShiftModifier, text="Z")
QKeyEvent(KeyPress, Key_Q, ShiftModifier, text="Q"
```

### In i3

```
Neovim API version compatible with 0 supported 5
Received unknown option "ext_hlstate" QVariant(bool, false)
QKeyEvent(KeyPress, Key_I, text="i")
QKeyEvent(KeyPress, Key_Super_L, ControlModifier)
QKeyEvent(KeyPress, Key_Escape, text="\u001B")
QKeyEvent(KeyPress, Key_Z, ShiftModifier, text="Z")
QKeyEvent(KeyPress, Key_Q, ShiftModifier, text="Q")
```

**Comment by hoatd** [comment]
It is the same to me in macOS. Cmd+V does not paste the system clipboard content but shows character 'V' instead. Cmd+v does its job pasting in nvim/terminal well

**Comment by jgehrig** [comment]
Sorry for the slow response, busy week...

Unfortunately, I still cannot reproduce the issue. Here is what I see:

1. Run neovim-qt `nvim-qt -- -u NORC`
2. Press `i`
3. Press `<Ctrl + Super>`
4. Press `<Esc>`
5. Press `Z`
6. Press `Q`

```
QKeyEvent(KeyPress, Key_I, text="i")
QKeyEvent(KeyPress, Key_Control, ControlModifier)
QKeyEvent(KeyPress, Key_Meta, ControlModifier|MetaModifier)
QKeyEvent(KeyPress, Key_Escape, text="\u001B")
QKeyEvent(KeyPress, Key_Shift, ShiftModifier)
QKeyEvent(KeyPress, Key_Z, ShiftModifier, text="Z")
QKeyEvent(KeyPress, Key_Q, ShiftModifier, text="Q")
```

Based on the information provided, I think we can rule out Desktop Manager or conflicting keybindings as a cause.

What version of Qt are you using? I am testing with Qt 5.12.3...

I also find `Key_Super_L` interesting... Why does my machine use Meta and yours Super? Does the L stand for "L key" or "Left"? Why does my machine send multiple events and yours only one?

**Comment by jgehrig** [comment]
@hoatd Your scenario is different, and not a bug :)

When running `nvim` through a terminal, Cmd + V is intercepted and never passed as a keystroke combination. Instead the contents of your clipboard is sent as if you typed it.

The behavior you are looking for is possible, but you will need to write a map in your vimrc. Such maps may conflict with the default bindings. For example, Ctrl+V is Visual Block Mode and not copy in Vim.

https://vim.fandom.com/wiki/Using_standard_editor_shortcuts_in_Vim

**Comment by beelzebielsk** [comment]
I'm seeing Qt 5.13.1 installed on my system.

<https://askubuntu.com/questions/19558/what-are-the-meta-super-and-hyper-keys>
<https://en.wikipedia.org/wiki/Super_key_(keyboard_button)#Linux_and_BSD>

Perhaps our initial assumption that `Meta` and `Super` are the same key is wrong. Not sure if that has an effect. I am pretty sure that the `_L` in `Key_Super_L` just means left.

If I run `xev`, which gives prints events on an X window to the terminal in which it is run, then pressing my `super` key gives me the key press event for `Super_L`.

Try pressing the key you've been pressing with `xev`. If it doesn't say `Super_L`, then I think you're pressing the wrong key.

<https://medium.com/@damko/a-simple-humble-but-comprehensive-guide-to-xkb-for-linux-6f1ad5e13450>
<https://askubuntu.com/questions/24916/how-do-i-remap-certain-keys-or-devices#24930>

According to these two, you can use `xmodmap` to change signal a key will send, and I'm using signal to mean things like `Meta` and `Super_L`.

On my system, the windows key is keycode 133, and I can change which signal is given by assigning a new signal to it using xmodmap. For example, if I wanted to assign `Home` to my windows key, I could write

    xmodmap -e "keycode 133 = Home" 

And it would work. So do me a favor and see what happens when you remap a key to be `Super_L`. If you need them, directions follow below.

**Drections to remap a key:**

1. Choose a key you'd like to remap to being `Super_L`. You can just choose your windows key on your keyboard if you have it.
2. Go to your terminal and run `xev`.
3. Press the key and look at the events that are printed in the terminal. Pay special attention to the following:
    - `keycode`: This will tell you the identity of the key being pressed, separate from what signal it is sending to X.
    - `keysym` : This will tell you the signal being sent to X by the key. Once you're done mapping your chosen key to `Super_L`,
      you'll want to map it back to being this. You could also restart
      your computer, I think.
4. Remap the signal of that key using the following: `xmodmap -e "keycode ${keycode} = Super_L"`. Careful, the keysym is case-sensitive.

Here's some sample output from `xev` when I press my `Super` key:

    KeyPress event, serial 34, synthetic NO, window 0x1200001,
        root 0x59e, subw 0x0, time 10415238, (-19,98), root:(944,152),
        state 0x0, keycode 133 (keysym 0xffeb, Super_L), same_screen YES,
        XLookupString gives 0 bytes:
        XmbLookupString gives 0 bytes:
        XFilterEvent returns: False

    KeyRelease event, serial 34, synthetic NO, window 0x1200001,
        root 0x59e, subw 0x0, time 10415550, (-19,98), root:(944,152),
        state 0x40, keycode 133 (keysym 0xffeb, Super_L), same_screen YES,
        XLookupString gives 0 bytes:
        XFilterEvent returns: False

**Comment by jgehrig** [comment]
Thanks for the detailed write-up. It was very helpful! :+1: 

Based on the ouput of xev, we are looking at the same key events from the perspective of X11. However, Qt/QApplication (for some unknown reason) is generating different QKeyEvent's when comparing our systems. Maybe this is a behavior change in 5.13? Maybe there is an overlooked setting?

Anyways... I can now reproduce your scenario in some sense. I have bound an arbitrary key "]" to the Super + Ctrl QKeyEvent your system generates. Here is the modification:
``` patch
diff -r 5556b921b590 src/gui/shell.cpp
--- a/src/gui/shell.cpp Wed Sep 18 18:06:11 2019 -0400
+++ b/src/gui/shell.cpp Wed Oct 23 01:59:51 2019 -0400
@@ -1065,6 +1065,11 @@
                return;
        }
 
+       QKeyEvent keyEvent{ QEvent::KeyPress, Qt::Key::Key_Super_L, Qt::ControlModifier};
+       if (ev->key() == 93) { // "]"
+               ev = &keyEvent;
+       }
+
        // conceal mouse pointer when typing
        if (m_mouseHide) {
                this->setCursor(Qt::BlankCursor);
@@ -1076,6 +1081,8 @@
                return;
        }
 
+       qDebug() << ev;
+
        m_nvim->api0()->vim_input(m_nvim->encode(inp));
        // FIXME: bytes might not be written, and need to be buffered
 }
```

It looks like this is a bug in `src/gui/input.cpp`. There probably needs to be an additional case called out somewhere in the following block:
https://github.com/equalsraf/neovim-qt/blob/9875617f1f471031ee48cb4f51757ba73e08833b/src/gui/input.cpp#L221-L234

My system returns at the first if, while your system returns at the final else block.

Once I have a patch, will you be able to test a fix?

**Comment by beelzebielsk** [comment]
Yes, sir.

**Comment by jgehrig** [comment]
Fixed in `v0.2.16`. Marking as closed.
